// @applitools/driver/dist/capabilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractCapabilitiesViewport = exports.extractCapabilitiesEnvironment = void 0;
function extractCapabilitiesEnvironment(capabilities) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    if (capabilities.capabilities)
        capabilities = capabilities.capabilities;
    const environment = {
        browserName: !capabilities.app && !capabilities.bundleId
            ? ((_a = capabilities.browserName) !== null && _a !== void 0 ? _a : (_b = capabilities.desired) === null || _b === void 0 ? void 0 : _b.browserName) || undefined
            : undefined,
        browserVersion: ((_c = capabilities.browserVersion) !== null && _c !== void 0 ? _c : capabilities.version) || undefined,
        platformName: ((_e = (_d = capabilities.platformName) !== null && _d !== void 0 ? _d : capabilities.platform) !== null && _e !== void 0 ? _e : (_f = capabilities.desired) === null || _f === void 0 ? void 0 : _f.platformName) || undefined,
        platformVersion: capabilities.platformVersion || undefined,
        isW3C: isW3C(capabilities),
        isMobile: isMobile(capabilities),
        isChrome: isChrome(capabilities),
        isECClient: Boolean(capabilities['applitools:isECClient']),
    };
    if (environment === null || environment === void 0 ? void 0 : environment.isMobile) {
        environment.deviceName = ((_h = (_g = capabilities.desired) === null || _g === void 0 ? void 0 : _g.deviceName) !== null && _h !== void 0 ? _h : capabilities.deviceName) || undefined;
        environment.isIOS = isIOS(capabilities);
        environment.isAndroid = isAndroid(capabilities);
        if (!environment.browserName) {
            environment.isNative = true;
        }
        else if (environment.isIOS && !/mobilesafari/i.test(capabilities.CFBundleIdentifier)) {
            environment.browserName = undefined;
            environment.isNative = true;
        }
        else {
            environment.isNative = false;
        }
    }
    return environment;
}
exports.extractCapabilitiesEnvironment = extractCapabilitiesEnvironment;
function extractCapabilitiesViewport(capabilities) {
    var _a, _b, _c, _d;
    if (capabilities.capabilities)
        capabilities = capabilities.capabilities;
    const viewport = {
        displaySize: extractDisplaySize(capabilities),
        orientation: (_b = ((_a = capabilities.deviceOrientation) !== null && _a !== void 0 ? _a : capabilities.orientation)) === null || _b === void 0 ? void 0 : _b.toLowerCase(),
        pixelRatio: capabilities.pixelRatio,
        statusBarSize: (_c = capabilities.statBarHeight) !== null && _c !== void 0 ? _c : (_d = capabilities.viewportRect) === null || _d === void 0 ? void 0 : _d.top,
    };
    if (viewport.displaySize && viewport.orientation && capabilities.viewportRect) {
        viewport.navigationBarSize =
            viewport.orientation === 'landscape'
                ? viewport.displaySize.width - (capabilities.viewportRect.left + capabilities.viewportRect.width)
                : viewport.displaySize.height - (capabilities.viewportRect.top + capabilities.viewportRect.height);
    }
    return viewport;
}
exports.extractCapabilitiesViewport = extractCapabilitiesViewport;
function isW3C(capabilities) {
    const isW3C = Boolean((capabilities.platformName || capabilities.browserVersion) &&
        (capabilities.platformVersion || capabilities.hasOwnProperty('setWindowRect')));
    return isW3C || isAppium(capabilities);
}
function isAppium(capabilities) {
    return (Boolean(capabilities.automationName || capabilities.deviceName || capabilities.appiumVersion) ||
        Object.keys(capabilities).some(cap => cap.startsWith('appium:')));
}
function isChrome(capabilities) {
    return Boolean(capabilities.chrome || capabilities['goog:chromeOptions']);
}
function _isFirefox(capabilities) {
    return capabilities.browserName === 'firefox' || Object.keys(capabilities).some(cap => cap.startsWith('moz:'));
}
function isMobile(capabilities) {
    var _a, _b;
    return (capabilities.browserName === '' ||
        ['ipad', 'iphone', 'android'].includes((_b = (_a = capabilities.browserName) === null || _a === void 0 ? void 0 : _a.toLowerCase()) !== null && _b !== void 0 ? _b : '') ||
        isAppium(capabilities));
}
function isIOS(capabilities) {
    return /iOS/i.test(capabilities.platformName) || /(iPad|iPhone)/i.test(capabilities.deviceName);
}
function isAndroid(capabilities) {
    return /Android/i.test(capabilities.platformName) || /Android/i.test(capabilities.browserName);
}
function extractDisplaySize(capabilities) {
    if (!capabilities.deviceScreenSize)
        return undefined;
    const [width, height] = capabilities.deviceScreenSize.split('x');
    if (Number.isNaN(Number(width)) || Number.isNaN(Number(height)))
        return undefined;
    return { width: Number(width), height: Number(height) };
}
